core-bundle/src/Controller/FrontendModule/TwoFactorController.php

```diff
@@ -0,0 +1,181 @@
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
+
+class TwoFactorController extends AbstractFrontendModuleController
+{
+    /**
+     * @var Translator
+     */
+    protected $translator;
+
+    /**
+     * @var RouterInterface
+     */
+    protected $router;
+
+    /**
+     * @var TokenStorage
+     */
+    protected $tokenStorage;
+
+    /**
+     * @var Authenticator
+     */
+    protected $authenticator;
+
+    /**
+     * @var AuthenticationUtils
+     */
+    protected $authenticationUtils;
+
+    /** @var PageModel */
+    protected $page;
+
+    public function __construct(Translator $translator, RouterInterface $router, TokenStorage $tokenStorage, Authenticator $authenticator, AuthenticationUtils $authenticationUtils)
+    {
+        $this->translator = $translator;
+        $this->router = $router;
+        $this->tokenStorage = $tokenStorage;
+        $this->authenticator = $authenticator;
+        $this->authenticationUtils = $authenticationUtils;
+    }
+
+    public function __invoke(Request $request, ModuleModel $model, string $section, array $classes = null, PageModel $page = null): Response
+    {
+        $this->page = $page;
+
+        if ($this->get('contao.routing.scope_matcher')->isFrontendRequest($request)) {
+            $this->page->loadDetails();
+        }
+
+        return parent::__invoke($request, $model, $section, $classes);
+    }
+
+    protected function getResponse(Template $template, ModuleModel $model, Request $request): Response
+    {
+        $user = $this->tokenStorage->getToken()->getUser();
+
+        if (!$user instanceof FrontendUser) {
+            return $template->getResponse();
+        }
+
+        /** @var PageModel $adapter */
+        $adapter = $this->get('contao.framework')->getAdapter(PageModel::class);
+
+        $redirectPage = $adapter->findByPk($model->jumpTo);
+        $return = $redirectPage instanceof PageModel ? $redirectPage->getAbsoluteUrl() : $this->page->getAbsoluteUrl();
+
+        $template->error = false;
+        $template->action = '';
+        $template->enforceTwoFactor = $this->page->enforceTwoFactor;
+        $template->targetPath = $return;
+
+        // Inform the user if 2FA is enforced
+        if ($this->page->enforceTwoFactor) {
+            $template->message = $this->translator->trans('MSC.twoFactorEnforced', [], 'contao_default');
+        }
+
+        if ((!$user->useTwoFactor && $this->page->enforceTwoFactor) || 'enable' === $request->get('2fa')) {
+            $this->enableTwoFactor($template, $request, $user, $return);
+        }
+
+        if (!$this->page->enforceTwoFactor && 'tl_two_factor_disable' === $request->request->get('FORM_SUBMIT')) {
+            $this->disableTwoFactor($user);
+        }
+
+        $template->isEnabled = $user->useTwoFactor;
+        $template->href = $this->router->generate('tl_page.'.$this->page->id, ['2fa' => 'enable']);
```

Comment:
Not sure this is currently a supported way to generate page URLs

Reply:
You're the man of the routing stuff, aren't you? 😎

Reply:
🤷‍♂ 
I know this _might_ work, but I don't think we should suggest this (yet), as we can simply use `$this->page->getAbsoluteUrl().'?2fa=enable'`

Reply:
Can you elaborate on why we should not use this yet?

Reply:
Because the route name is *internal*. There are multiple routes per page.
Also, the cmf-router allows to route an object, so we could support `$router->generate($pageModel)` or `$router->generate($newsModel)`, but we need to add support for that first by implementing some stuff. I think if we want to use the actual router, we should use objects and not generate an *internal* route name.

Reply:
Fixed in ff9e7e28d.